syntax = "proto3";

import "agent.proto";

package gaiachain;

message Entity {
    string id = 1;

    uint64 timestamp = 2;
    Location location = 21;
    Agent user = 22;

    enum PackageType {
        ZERO_TYPE = 0; // zero valued enums are not shown in details
        HARVEST = 1;
        SAC = 2;
        LOT = 3;
    }
    // Package type of entity
    PackageType package_type = 3;

    // step 1
    uint64 harvest_date = 4;
    string parcel = 5;

    // step 2
    uint64 breaking_date = 6;
    uint64 end_fermentation_date = 7;
    uint64 beans_volume = 8;

    // step 3
    uint64 reception_date = 9;
    uint64 transport_date = 10; // also used in step 6
    string buyer = 11;

    repeated Entity entities = 12;

    enum Status {
        ZERO_STATUS = 0; // zero valued enums are not shown in details
        HARVESTING = 1;
        GRAIN_PROCESSING = 2;
        SECTION_RECEPTION = 3;
        BAGGING = 4;
        LOT_CREATION = 5;
        WAREHOUSE_TRANSPORT = 6;
        EXPORT_RECEPTION = 7;
    }
    Status status = 14;

    // step 4
    string sac_id = 15;
    uint64 weight = 16; // also used in step 7

    // step 5
    string lot_id = 17;
    string notes = 18;

    // step 6
    string transporter = 19;
    string destination = 20;

}

message Location {
  double lat = 1;
  double long = 2;
}

message EntityBatch {
    // Current owner public_key
    // string owner_public_key = 1;

    // Entities in bath.
    repeated Entity entities = 1;

    bool finished = 3;

    string id = 4;
}

